ethers-middleware/src/multicall.rs

```diff
@@ -0,0 +1,386 @@
+use thiserror::Error;
+
+use tokio::sync::{mpsc, oneshot};
+
+type MulticallResult<M> = Result<Token, Arc<MulticallError<M>>>;
+type MulticallRequest<M> = (ContractCall<M, Token>, oneshot::Sender<MulticallResult<M>>);
+
+#[derive(Debug)]
+/// Processor for multicall middleware requests
+pub struct MulticallProcessor<M: Middleware> {
+    inner: Arc<M>,
+    multicall_address: Option<Address>,
+    max_batch_size: usize,
+    rx: mpsc::Receiver<MulticallRequest<M>>,
+}
+
+#[derive(Debug, Clone)]
+/// Middleware used for transparently leveraging multicall functionality
+pub struct MulticallMiddleware<M: Middleware> {
+    inner: Arc<M>,
+    contracts: Vec<BaseContract>,
+    tx: mpsc::Sender<MulticallRequest<M>>,
+}
+
+#[derive(Error, Debug)]
+/// Thrown when an error happens at the Multicall middleware
+pub enum MulticallMiddlewareError<M: Middleware> {
+    /// Thrown when the internal middleware errors
+    #[error("{0}")]
+    MiddlewareError(M::Error),
+    /// Thrown when the internal multicall errors
+    #[error(transparent)]
+    MulticallError(#[from] Arc<MulticallError<M>>),
+    /// Thrown when the processor isn't running
+    #[error("Processor is not running")]
+    ProcessorNotRunning,
+}
+
+impl<M: Middleware> MiddlewareError for MulticallMiddlewareError<M> {
+    type Inner = M::Error;
+
+    fn from_err(src: M::Error) -> Self {
+        MulticallMiddlewareError::MiddlewareError(src)
+    }
+
+    fn as_inner(&self) -> Option<&Self::Inner> {
+        match self {
+            MulticallMiddlewareError::MiddlewareError(e) => Some(e),
+            MulticallMiddlewareError::MulticallError(e) => e.as_middleware_error(),
+            MulticallMiddlewareError::ProcessorNotRunning => None,
+        }
+    }
+}
+
+impl<M> MulticallProcessor<M>
+where
+    M: Middleware,
+{
+    /// Should be run in a separate task to process requests
+    pub async fn run(mut self) {
+        let mut multicall: Multicall<M> =
+            Multicall::new(self.inner, self.multicall_address).await.unwrap();
+
+        loop {
+            let mut requests = Vec::new();
+
+            // wait for the first request
+            match self.rx.recv().await {
+                Some(request) => requests.push(request),
+                None => break,
+            }
+
+            // attempt to batch more requests, up to the max batch size
+            while requests.len() < self.max_batch_size {
+                match self.rx.try_recv() {
+                    Ok(request) => requests.push(request),
+
+                    // For both errors (Disconnected and Empty), the correct action
+                    // is to process the items.  If the error was Disconnected, on
+                    // the next iteration rx.recv().await will be None and we'll
+                    // break from the outer loop anyway.
+                    Err(_) => break,
+                }
+            }
+
+            let (calls, callbacks): (Vec<_>, Vec<_>) = requests.into_iter().unzip();
+
```

Comment:
do we need any special handling if there's only one request?

Reply:
we don't *need* special handling but its a minor optimization?